doc/designs/quic-design/quic-concurrency.md

```diff
@@ -0,0 +1,412 @@
+fully abstracts the QUIC core objects such as `QUIC_ENGINE` or `QUIC_CHANNEL` so
+that the APL never sees them.
+
+The caller retrieves a pipe handle using `ossl_cml_get_pipe`. This function
+retrieves a pipe based on two values:
+
+  - a CML pipe class;
+  - a CML *selector*.
+
+The CML selector is a tagged union structure which specifies what pipe is to be
+retrieved. Abstractly, examples of selectors include:
+
+```text
+    Domain      ()
+    Listener    (listener_id: uint)
+    Conn        (conn_id:     uint)
+    Stream      (conn_id:     uint, stream_id: u64)
+```
+
+In other words, the CML selector selects the “object” to retrieve a pipe from.
+
+The CML pipe class is one of the following values:
+
+- Request
+- Notification
+- App Send
+- App Recv
+
+The pipe classes available for a given selector vary. For example, the “App
+Send” and “App Recv” pipes only exist on a stream, so it is invalid to request
+such a pipe in conjunction with a different type of selector.
+
+The “Request” and “App Send” classes expose send-only streams, and the
+“Notification” and “App Recv” classes expose receive-only streams.
+
+For any given CML selector, the Request pipe is used to send serialized commands
+for asynchronous processing in relation to the entity selected by that selector.
+Conversely, the Notification pipe returns asynchronous notifications. These
+could be in relation to a previous Command (e.g. indicating whether a command
+succeeded), or unprompted notifications about other events.
+
+The underlying pattern here is that there is a bidirectional channel for control
+messages, and a bidirectional channel for application data, both comprised of
+two unidirectional pipes in turn.
+
+Pipe handles are stable for as long as the pipe they reference exists, so an APL
+object can cache a pipe handle if desired.
+
+All CML methods are thread safe. The CML implementation handles any necessary
+locking (if any) internally.
+
+The `ossl_cml_write_available` and `ossl_cml_read_available` calls determine the
+number of bytes which can currently be written to a send-only pipe, or read from
+a receive-only pipe, respectively.
+
+**Race conditions.** Because these are separate calls to `ossl_cml_write` and
+`ossl_cml_read`, the values returned by these functions may become out of date
+before the caller has a chance to read `ossl_cml_write` or `ossl_cml_read`.
+However, such changes are guaranteed to be monotonically in favour of the
+caller; for example, the value returned by `ossl_cml_write_available` will only
+ever increase asynchronously (and only decrease as a result of an
+`ossl_cml_write` call). Conversely, the value returned by
+`ossl_cml_read_available` will only ever increase asynchronously (and only
+decrease as a result of an `ossl_cml_read` call). Assuming that only one thread
+makes calls to CML functions at a given time *for a given pipe*, this therefore
+poses no issue for callers.
+
```

Comment:
Is  "no issue for callers" really accurate here?  It makes sense if the threading constraints are met, but it seems the entire point of this concurrency model is to allow multiple threads to do I/O on a given quic connection/stream in parallell.

Reply:
> Is "no issue for callers" really accurate here? It makes sense if the threading constraints are met, but it seems the entire point of this concurrency model is to allow multiple threads to do I/O on a given quic connection/stream in parallell.

UDP is not required to preserve send order packets on the receiving side. It is up to the caller to determine the packet order and rearrange as needed. A typical example of this is where the routing tables are being updated during transmission - assuming there is a sequencing mechanism in the packets implemented by the application, and that a recvfrom() is always pending so zero packet loss due to timing.

1. sendto(seq 1) issued in peer1
2. recvfrom(seq 1) completes in peer2
3. sendto(seq 2) issued in peer1
4. Routing table update shortening the route between peer 1 and peer 2
5. sendto(seq 3) issue in peer 1
6. recvfrom(seq 3) completes in peer2
7. recvfrom(seq 2) completes in peer2

This may be irrelevant to OpenSSL, and if so, I apologize.

Reply:
UDP has no ordering requirements, but QUIC does.  RFC 9000 section 2.2 indicates that stream frames within a connection contain an offset field indicating the ordering of data within the stream.  The core of the QUIC code receives this data from the underlying socket and queues it for availability to the application up to the point where there are no gaps in the stream (see ossl_quic_rstream_queue_data for an example of where this is done).  Applications reading quic streams (via quic_read or one of the proposed concurrency models) will only read byte streams in order.